fs/reader/reader.go

```diff
@@ -227,11 +227,10 @@ func (sf *file) ReadAt(p []byte, offset int64) (int, error) {
 	commonmetrics.IncOperationCount(commonmetrics.OnDemandRemoteRegistryFetchCount, sf.gr.layerSha) // increment the number of on demand file fetches from remote registry
 	sf.gr.setLastReadTime(time.Now())
 
-	contents, err := io.ReadAll(r)
+	n, err := io.ReadFull(r, p[0:expectedSize])
 	if err != nil && err != io.EOF {
 		return 0, err
 	}
-	n := copy(p, contents)
 	if soci.FileSize(n) != expectedSize {
```

Comment:
Can we combine this with the previous error case?

The possible returns from `io.ReadFull` are (let `buflen == len(p[0:expectedSize]`):
1) n == buflen, err == nil
2) n == 0 err == `io.EOF`
3) 0 < n < buflen, err == `io.ErrUnexpectedEOF`
4) n == ?, err == other read error

Can we just combine these into:
```
n, err := io.ReadFull(r, p[0:expectedSize])
if err != nil {
        return 0, fmt.Errorf("unexpected copied data size for on-demand fetch. read = %d, expected = %d: %w", n, expectedSize, err)
}
```

Reply:
Fair point - could we guard for that with a precondition?
```
if len(p) == 0 {
    return 0, nil
}
```

Reply:
Done